doc/developer/design/20210412_postgres_sources.md

```diff
@@ -0,0 +1,210 @@
+  replication slot is a bit complicated so the MVP will assume the initial
+  connection will remain open. Lifting this assumption should be prioritized
+  after the MVP.
+
+
+## Description
+
+### Terminology
+
+Logical replication involves a few different postgres objects that work together
+to move data from one postgres instance to another.
+
+1. **Logical replication**: This is a special mode of postgres streaming
+   replication where data is streamed transaction by transaction in commit
+   order. This is in contrast with normal replication where binary segments of
+   the WAL are sent to the downstream database.
+2. **Decoding plugin**: An component that runs inside the upstream database that
+   is responsible for reading the raw WAL segments and reconstructing the
+   logical transaction messages before sending them over to the logical
+   replication receiver.
+3. **pgoutput**: The specific decoder that postgres uses when logical
+   replication is used. There are a lot of third-party plugins available but
+   pgoutput is offered by postgres itself.
+4. **PUBLICATION**: It is an object that tells the upstream database which
+   tables are to be published over a replication stream. This can be an explicit
+   list of table names or a special "all tables" publication that gets
+   everything.
+5. **replication slot**: It is an object that keeps track of how far along a
+   replica is with its stream consumption. Normally the upstream database has
+   one such object for each downstream replica. Each replica periodically sends
+   updates with the WAL offset up to which it has consumed which then lets the
+   upstream database safely clean up segments from disk.
+
+### Replication mechanism
+
+In order to replicate a set of tables from an upstream database a replica needs
+to orchestrate two phases, making sure that there is no data loss during the
+handover. The first phase involves the initial table sync which is done with
+normal `COPY` bulk transfer commands. Once the replica has the initial snapshot
+of a table it can switch to reading updates from a replication slot in order to
+keep its table synced.
+
+Ensuring that there is no gap between finishing the initial COPY and starting
+reading from the replication slot is done by synchronizing the database snapshot
+at which the `COPY` is performed with the starting point of the replication
+slot. The simplest way to do this is by using the following query structure:
+
+```sql
+-- Initial sync phase
+BEGIN READ ONLY ISOLATION LEVEL REPEATABLE READ
+CREATE_REPLICATION_SLOT <foo> LOGICAL pgoutput USE_SNAPSHOT;
+COPY `table` TO STDOUT;
+COMMIT
+
+-- Streaming phase
+START_REPLICATION SLOT <foo>
+```
+
+By creating the replication slot as the first statement in a transaction and
+setting the `USE_SNAPSHOT` option we ensure that the transaction and the slot
+are at the exact same point. The existence of the slot will prevent any WAL
```

Comment:
> The existence of the slot will prevent any WAL segments from being cleaned up...

Since the slot will outlive the `COPY` process, does this mean that WAL segments will not be able to be cleaned up for the entirety of the replication process? (As long as a Postgres source in Materialize is live?) Or does something about the slot change when the `COPY` step has completed?

Reply:
I think I answered my own question from [the Postgres docs!](https://www.postgresql.org/docs/13/warm-standby.html#STREAMING-REPLICATION-SLOTS) 
> Replication slots provide an automated way to ensure that the master does not remove WAL segments until they have been received by all standbys, and that the master does not remove rows which could cause a recovery conflict even when the standby is disconnected.

Reply:
The specific mechanism is through the standby status update messages in the protocol that let the upstream know up to what LSN is safe to clean up